### ISSUE TITLE
如何监控和诊断大模型训练过程？

### ISSUE THREAD

**Issue by lihe** [lihe/MyGitBlog#38]
# **一、监控分三层**

我们把监控分为三类：

```
数值健康
计算效率
系统稳定
```

------

# **二、第一层：数值健康监控（模型是否在正常学习）**

这是最关键的一层。

------

## **1️⃣ Loss 曲线（不只是看下降）**

你要看：

- 是否单调下降？
- 是否突然 spike？
- 训练/验证是否分离？

异常模式：

| **现象**       | **可能原因**          |
| -------------- | --------------------- |
| loss 不动      | 学习率太小 / 梯度为0  |
| loss 爆炸      | 学习率太大 / 梯度爆炸 |
| 训练降、验证升 | 过拟合                |

------

## **2️⃣ 梯度范数（grad norm）**

这是比 loss 更早的预警指标。

观察：

$\|\nabla \theta\|_2$

异常模式：

- 逐步变小 → 梯度消失
- 突然变大 → 即将 NaN

建议：

记录：

```
torch.nn.utils.clip_grad_norm_
```

并画曲线。

------

## **3️⃣ 参数分布**

看：

- 权重均值
- 权重方差
- 是否偏移

异常：

- 全部集中到 0
- 方差爆炸

说明初始化或优化器有问题。

------

## **4️⃣ 注意力矩阵分布**

观察：

- 是否极端稀疏？
- 是否全接近均匀？

如果注意力 collapse：

模型退化。

------

# **三、第二层：训练效率监控（算力是否浪费）**

很多人忽略这部分。

------

## **1️⃣ GPU 利用率**

目标：

> 80%

如果只有 40%：

说明：

- 数据加载慢
- 通信阻塞
- micro-batch 太小

------

## **2️⃣ 显存占用曲线**

关注：

- 是否逐步上升（内存泄漏）
- 是否频繁 OOM

------

## **3️⃣ Tokens/sec**

真正衡量效率的指标。

例如：

```
1.2M tokens/sec
```

比 loss 更重要。

------

## **4️⃣ MFU（Model FLOPs Utilization）**

计算：

$\text{MFU} = \frac{\text{实际 FLOPs}}{\text{理论 FLOPs}}$

低于 40%：

系统有严重问题。

------

# **四、第三层：系统级监控**

------

## **1️⃣ 通信时间占比**

如果：

AllReduce 时间 > 计算时间

说明：

- 带宽瓶颈
- 拓扑设计不合理

------

## **2️⃣ NCCL 错误监控**

分布式常见问题：

- 死锁
- timeout
- rank 不同步

------

## **3️⃣ 节点健康**

监控：

- GPU 温度
- CPU IO
- 网络带宽

------

# **五、典型问题诊断路径**

我给你一个工业级排查逻辑。

------

## **情况 1：Loss 不下降**

检查顺序：

1. 学习率
2. 数据是否 shuffle
3. 梯度是否为 0
4. 是否错误加载 checkpoint

------

## **情况 2：突然 NaN**

检查：

1. grad norm 是否暴涨
2. 是否未启用 gradient scaling
3. 是否权重溢出

------

## **情况 3：GPU 利用率低**

检查：

1. DataLoader 是否卡
2. 是否同步阻塞
3. batch 是否太小
4. 是否 checkpoint 过多

------

# **六、真正高级监控（大厂做法）**

------

## **1️⃣ 自动异常检测**

- 监控 loss 曲率
- 监控梯度变化率

一旦异常：

自动降低学习率。

------

## **2️⃣ 在线验证**

每 N steps：

做小规模 eval。

防止：

一周后才发现模型崩溃。

------

## **3️⃣ 训练可恢复性**

- 定期保存 optimizer state
- 异步 checkpoint
- 自动重启机制

------

# **七、工业训练监控框架结构**

```
训练节点
   ↓
TensorBoard + Prometheus
   ↓
Grafana Dashboard
   ↓
异常报警（Slack/邮件）
```

------

# **八、监控优先级排序**

如果只能选 5 个指标：

1. Loss
2. Grad Norm
3. GPU Util
4. Tokens/sec
5. 显存

------

# **九、为什么大模型更需要监控？**

因为：

- 训练时间长（几周）
- 成本高（百万美元）
- 问题传播慢

一个小错误可能 3 天后才显现。

------

# **十、面试级总结回答**

如果问：

> 如何监控大模型训练？

可以回答：

> 需要从数值健康、计算效率和系统稳定三个层面监控，包括loss、梯度范数、参数分布、GPU利用率、通信时间和tokens/sec等关键指标，并配合自动异常检测与恢复机制，确保长时间训练的稳定性。

------

# **十一、终极理解**

大模型训练本质是：

$\text{长期稳定数值迭代系统}$

不是简单的：

“跑一个 PyTorch 脚本”。

------
